### PR TITLE
Add skeleton for hugr dialect

### PR DESCRIPTION
**Pull request by doug-q**
This looks large, and it is, but it's mostly cmake and boilerplate.

I have disabled zlib and terminfo in the LLVM derivation, as an alternative to working out why cmake can't find them when run in the nix shell.

Tests, which now include:
* checking the mlir derivation
* In both the build and release trees:
  * Checking `hugr-mlir-opt` can handle the skeleton dialect
  * Checking the C api bindings work 
  * Checking the python bindings work
* nix linting
* C++ linting
are all expected to pass. I've not seen green CI yet, so I may yet have to push small fixes.

Docs, which consist only of generated documentation, should be verified in an install tree. 

@aborgna-q Can you review?

### CODE REVIEW

File: mlir/include/hugr-mlir/IR/HugrDialect.td [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+def Hugr_Dialect : Dialect {
+    let name = "hugr";
+    let summary = "A dialect representing hugr graphs";
+    let description = [{
+        A dialect intended to facilitate importing and exporting [hugr](https://github.com/CQCL-DEV/hugr/blob/main/specification/hugr.md)
+        through mlir.
+
+        We model the full `hugr` type system in mlir types. We intend to
+        delegate to the `hugr` crate whenever possible through the companion
```
Comment (aborgna-q) [inline mlir/include/hugr-mlir/IR/HugrDialect.td:20]: The (to be) published crate is called `quantinuum-hugr`.

File: README.md
Comment (aborgna-q) [inline README.md:1]: It'd be nice to have a section after `### Build` on running the py tests.
(or alternatively in a DEVELOPMENT.md file)

**Review by aborgna-q — APPROVED** [review]
LGTM for my limited knowledge in mlir 👍

**Comment by doug-q** [comment]
Thanks @aborgna-q , I will make the remaining fixes in a followup as discussed

**Pull request merged**